Organize multimedia paths: rename

orangejenny opened the pull request:
Spec: https://docs.google.com/document/d/1fVzCgEw3-deesJqlHtBqy-CrvNvHhylmOAwODOQSdvc/edit

Followup for https://github.com/dimagi/commcare-hq/pull/22693 and https://github.com/dimagi/commcare-hq/pull/22745

This PR finishes the basic functionality: doing the actual renames specified in the uploaded file.

@mkangia / @emord 

<img width="1122" alt="screen shot 2018-12-18 at 11 58 38 pm" src="https://user-images.githubusercontent.com/1486591/50199806-f245ac80-0320-11e9-805d-b982402600bb.png">

[QA ticket](https://manage.dimagi.com/default.asp?287354)

orangejenny committed:
Added support for saving module icon only

orangejenny committed:
Added success messaging to page when update finishes.

orangejenny committed:
Added spinner behavior

orangejenny committed:
Added warning if requested renames are not completed.

orangejenny committed:
Added support for renaming questions in forms

orangejenny committed:
Added support for module menu audio, form menu icon and audio

orangejenny committed:
Moved success message generation to server side.

Stops displaying "0 items updated" messages, and order the messages
according to module/form order in the app.

orangejenny committed:
Cleaned up main update loop

Made this one outer loop over the list of paths.
Moved xfrom-specific logic into xform.py

orangejenny committed:
Added support for non-menu module multimedia

orangejenny committed:
Fixed bug where case list lookup image didn't show in multimedia reference page

orangejenny committed:
Fixed lint error

mkangia commented on corehq/apps/hqmedia/views.py:
```diff
@@ -270,10 +272,101 @@ def update_multimedia_paths(request, domain, app_id):
+            })
+
+        # Master list of old_path => new_path
+        paths = {row[0]: interpolate_media_path(row[1]) for row in rows}
```
probably a miss in my knowledge but why can't we ask the user to give us path which does not need us to interpolate?

orangejenny replied:
It does throw a warning, but as long as we can interpolate for them, we might as well do it.

mkangia replied:
if i am user and i am editing the paths, i would remove the extra `jr://` thing in beginning and then also update the path or i would just modify the path? i am assuming that the users start with the `jr://` path.

orangejenny replied:
As the user, you can do either. If you enter `path/to/thing`, `interpolate_media_path` will transform that to `jr://path/to/thing`. But if you enter `jr://path/to/thing`, `interpolate_media_path` won't do anything.

mkangia commented on corehq/apps/hqmedia/views.py:
```diff
@@ -270,10 +272,101 @@ def update_multimedia_paths(request, domain, app_id):
+        success_counts = defaultdict(lambda: 0)
+        xforms_by_id = defaultdict(lambda: None)    # cache forms rather than parsing repeatedly
+        dirty_xform_ids = set()
+        for old_path, new_path in six.iteritems(paths):
```
instead of iterating each pair, why cant be pass everything to each module/forms etc and them update all at once?

orangejenny replied:
The module/form would need to iterate over all of the paths, though. I'd rather do the list processing here, keeping all of the "bulk" logic in one place.

mkangia commented on corehq/apps/hqmedia/views.py:
```diff
@@ -270,10 +272,101 @@ def update_multimedia_paths(request, domain, app_id):
             'error': _("Could not find file. Please re-upload file.")
         })
 
-    # TODO: update paths
```
Can you break this down into small logical units/methods? Would help better understand and naming would give some documentation by default.

orangejenny replied:
yeah, I'll take a look at this

orangejenny replied:
Did a bunch of reorganization, starting with 5ec9ead6aff652d4ab3d856745f1d1299d6f6f64 The overall change is to move all of the renaming logic into hqmedia.models, and to break HQMediaMixin into smaller classes: one for apps, one for modules, and one for forms.

orangejenny commented on corehq/apps/hqmedia/static/hqmedia/js/manage_paths_main.js:
```diff
@@ -5,12 +5,17 @@ hqDefine("hqmedia/js/manage_paths_main", function () {
 
+        // Controls for disabling, etc in UI
+        self.isValidating = ko.observable(false);
+        self.isUpdating = ko.observable(false);
```
Going to leave it as is - having them as two separate flags lets me treat the two buttons differently (so only the "working" button has a spinner).

mkangia replied:
Can the two separate buttons be clicked together?

mkangia replied:
offlined. no actions needed.

orangejenny committed:
Renamed success flag to complete

orangejenny commented on corehq/apps/hqmedia/views.py:
```diff
@@ -270,10 +272,101 @@ def update_multimedia_paths(request, domain, app_id):
+        (errors, warnings) = validate_multimedia_paths_rows(app, rows)
+        if len(errors):
+            return json_response({
+                'success': 1,
```
Good idea - b1f4a4bbe17

orangejenny committed:
Extracted update_multimedia_paths into view_helpers

orangejenny committed:
Added test

orangejenny committed:
Fixed extraction update_multimedia_paths into view_helpers

orangejenny committed:
Fixed lint errors

orangejenny committed:
Fixed lint error

orangejenny committed:
Renamed HQMediaMixin and added MediaMixin

orangejenny committed:
Moved case list icons from app's all_media to ModuleMediaMixin

orangejenny committed:
Moved module and form menu media from app's all_media to Module/FormMediaMixin

orangejenny committed:
Moved form media from app's all_media to FormMediaMixin

orangejenny committed:
Moved remaining detail media to ModuleMediaMixin

orangejenny committed:
Consolidated code to get menu media

orangejenny committed:
Moved app-level all_media into ApplicationMediaMixin

orangejenny committed:
Moved all_media-related methods to MediaMixin

orangejenny committed:
Combined MediaControllerMixin and ApplicationMediaMixin

Decided it was simpler to have one application-level media class.
I'd like to move more of the logic in ApplicationMediaMixin out into
ModuleMixin and FormMixin, and DRY it up with the logic in all_media,
but I'm calling that out of scope for this PR.

orangejenny committed:
Fixed lack of error when upload 500s

orangejenny committed:
Moved rename logic into hqmedia.models

orangejenny committed:
Fixed lint errors

orangejenny commented:
@mkangia This is ready for re-review.

orangejenny committed:
Fixed import error on multimedia download

mkangia commented on corehq/apps/hqmedia/models.py:
```diff
@@ -540,10 +552,198 @@ def _log_media_deletion(app, deleted_media):
+        }
+
+    @memoized
+    def memoized_xform(self):
```
Why is it called `memoized_xform` and not just `xform`?

orangejenny replied:
I wanted to be explicit, don't want someone to see `form.xform` in the code and not realize it's memoized.

mkangia replied:
Okay. We do memoize properties at lots of places but necessarily not name that like that. Fine to keep it like this if you think adds value.

mkangia commented on corehq/apps/hqmedia/models.py:
```diff
@@ -540,10 +552,198 @@ def _log_media_deletion(app, deleted_media):
+        kwargs = self.get_media_ref_kwargs()
+        media = []
+
+        media.extend(self.menu_media(self))
```
whats the reason for using extend and not just a straight assignment?

orangejenny replied:
Not deliberate. Updated.

mkangia reviewed (commented):
took another pass.

orangejenny committed:
Added error if old and new paths are the same

orangejenny committed:
Updated workflow and file format.

- Changed upload UI to have a single button instead of one to validate and a second to upload
- Added header and blank "New Path" column to download so it can be editied & re-uploaded
- Removed error on number of columns, since users are likely to include usages columns on re-upload

orangejenny commented on corehq/apps/hqmedia/models.py:
```diff
@@ -540,10 +552,198 @@ def _log_media_deletion(app, deleted_media):
+        return set([m.path for m in self.all_media()])
+
+    @memoized
+    def get_all_paths_of_type(self, media_class_name):
```
sure, will update

orangejenny commented on corehq/apps/hqmedia/models.py:
```diff
@@ -540,10 +552,198 @@ def _log_media_deletion(app, deleted_media):
+            Convenience method. Gets the ApplicationMediaReference for a menu that's
+            associated with this MediaMixin (which may be self or a different attribute)
+        """
+        kwargs = self.get_media_ref_kwargs()
```
yes, thank you

orangejenny replied:
oops - no, that's not right. `get_media_ref_kwargs` is a function of `MediaMixin`. `self` is going to be a `MediaMixin`, but `menu` might not be. `menu` only has to implement `NavMenuItemMediaMixin`. I added a comment to `menu_media` while responding to other feedback, hopefully that'll clear it up a bit.

orangejenny commented on corehq/apps/hqmedia/models.py:
```diff
@@ -540,10 +552,198 @@ def _log_media_deletion(app, deleted_media):
+        media.extend(self.menu_media(self))
+
+        if self.case_list_form.form_id:
+            media.extend(self.menu_media(self.case_list_form))
```
ugh, yeah, now that we're calling modules "menus" in the ui this is a bit confusing. I'm not keen to rename NavMenuItemMediaMixin, but yeah, will add a comment to the definition of `menu_media`.

orangejenny commented on corehq/apps/hqmedia/models.py:
```diff
@@ -540,10 +552,198 @@ def _log_media_deletion(app, deleted_media):
+        if self.case_list_form.form_id:
+            media.extend(self.menu_media(self.case_list_form))
+
+        if hasattr(self, 'case_list') and self.case_list.show:
```
added comments noting what features these refer to, let me know if that's not what you meant

orangejenny commented on corehq/apps/hqmedia/models.py:
```diff
@@ -540,10 +552,198 @@ def _log_media_deletion(app, deleted_media):
+        }
+
+    def all_media(self):
+        kwargs = self.get_media_ref_kwargs()
```
Yep, that's correct, and it's not great. I think addressing it is a big undertaking, given how many ways we store media (top-level attribute on the module or form, attribute of another attribute, etc., not to mention media that's stored in xforms). I'll keep thinking about it, but not going to make changes to this PR.

New media also needs to be added both to `all_media` and `rename_media` which I think is the core weakness of this particular PR (although it's somewhat mitigated by the warning that pops up if after the rename there are any reference left to the old name).

orangejenny commented on corehq/apps/hqmedia/tests/test_manage_paths.py:
```diff
@@ -72,27 +79,118 @@ def test_paths_download(self, mock):
+            (self._get_form_media('en'), 'jr://file/tulip.png'),      # 3. valid
+            (self._get_form_media('en'), 'jr://file/hyacinth.png'),   # 4. error & warning: dupe old and new paths
+            (self._get_form_media('fra'), 'jr://file/hyacinth.png'),  # 5. warning: dupe new path
+            (self._get_form_media('en'), self._get_form_media('en')), # 6. error: same old and new path
```
ignoring

orangejenny committed:
Added comments and did minor cleanup

orangejenny committed:
Added warning on changing file extension

emord commented on corehq/apps/app_manager/models.py:
```diff
@@ -1710,7 +1715,7 @@ def custom_icon(self):
 
 
-class Form(IndexedFormBase, NavMenuItemMediaMixin):
+class Form(IndexedFormBase, FormMediaMixin, NavMenuItemMediaMixin):
```
Would these media mixins be better on the `IndexedFormBase` or `FormBase` classes? Actually I'm not even sure why we need both of those base classes in general

orangejenny replied:
Now that you mention it, I don't know why we have both `FormBase` and `IndexedFormBase` either, or why `NavMenuItemMediaMixin` isn't mixed in to one of those instead of where it is. I'll take a stab at simplifying the class hierarchy, but will do it in a followup PR.

emord commented on corehq/apps/hqmedia/urls.py:
```diff
@@ -32,7 +31,6 @@
     url(r'^upload/$', BulkUploadMultimediaView.as_view(), name=BulkUploadMultimediaView.urlname),
     url(r'^paths/$', ManageMultimediaPathsView.as_view(), name=ManageMultimediaPathsView.urlname),
     url(r'^paths/download/$', download_multimedia_paths, name='download_multimedia_paths'),
-    url(r'^paths/validate/$', validate_multimedia_paths, name='validate_multimedia_paths'),
```
nit: removing a URL/view like this isn't safe. scenario:

1. user opens the page in a background tab that links to path validation
2. we deploy this PR
3. that user goes to that tab, to validate paths. 
4. They are greeted by either a 4xx or 500 error

I don't know enough about this workflow to say how big of a risk it is, so not requiring changes, will let you decide

orangejenny replied:
Fair point in general. Not worried about it for this PR because no one is using this feature - I released it in pieces because it was originally on a tight timeline and having it partially released seemed potentially useful.

mkangia commented:
🎉 🎆
